src/responder.ts

```diff
@@ -150,83 +176,329 @@ export class EthereumDedicatedResponder extends EthereumResponder {
-     * @param ethereumResponse The IEthereumResponse containing what needs to be submitted.
      * @param [confirmationsRequired=40] The number of confirmations required before a transaction is trusted.
      * @param [maxAttempts=10] The maximum number of retries before the Responder will give up.
      */
     constructor(
-        readonly signer: ethers.Signer,
-        public readonly appointmentId: string,
-        public readonly ethereumResponse: IEthereumResponse,
-
+        signer: ethers.Signer,
         public readonly confirmationsRequired = 40,
         private readonly maxAttempts: number = 10
     ) {
-        super(signer, appointmentId, ethereumResponse);
+        super(signer);
+    }
 
-        signer.provider.on("block", this.updateLastBlockTime);
+    private setup() {
+        this.locked = true;
+        this.signer.provider.on("block", this.newBlockReceived.bind(this));
     }
 
     /**
      * Release any resources used by this instance.
      */
     private destroy() {
-        this.signer.provider.removeListener("block", this.updateLastBlockTime);
+        this.signer.provider.removeListener("block", this.newBlockReceived);
+        this.locked = false;
     }
 
-    private updateLastBlockTime() {
+    private newBlockReceived(blockNumber: number) {
+        this.lastBlockNumberSeen = blockNumber;
         this.timeLastBlockReceived = Date.now();
     }
 
-    public async respond() {
+    public async startResponse(appointmentId: string, responseData: IEthereumResponseData) {
+        if (this.locked) {
+            throw new Error("This responder can ony handle one response at a time."); // TODO: more specific Error type?
+        }
+
+        const responseFlow = new EthereumResponseFlow(appointmentId, responseData);
+
+        this.setup();
+
+        const signerAddress = await promiseTimeout(
+            this.signer.getAddress(),
+            EthereumDedicatedResponder.WAIT_TIME_FOR_PROVIDER_RESPONSE
+        );
+
+        // Get the current nonce to be used
+        this.nonce = await promiseTimeout(
+            this.signer.provider.getTransactionCount(signerAddress),
+            EthereumDedicatedResponder.WAIT_TIME_FOR_PROVIDER_RESPONSE
+        );
+
+        // Get the initial gas price
+        this.gasPrice = await promiseTimeout(
+            this.signer.provider.getGasPrice(),
+            EthereumDedicatedResponder.WAIT_TIME_FOR_PROVIDER_RESPONSE
+        );
+
         while (this.mAttemptsDone < this.maxAttempts) {
             this.mAttemptsDone++;
             try {
                 // Try to call submitStateFunction, but timeout with an error if
-                // there is no response for 30 seconds.
-                const tx = await promiseTimeout(this.submitStateFunction(), 30000);
+                // there is no response for WAIT_TIME_FOR_PROVIDER_RESPONSE ms.
+                const tx = await promiseTimeout(
```

Comment:
I think this section should be split into three different parts. 1. Put tx into pending. 2. Mine tx. 3. Confim tx.

I think you should create a new class eg. TransactionMiner, with these three methods on it and call it in here. This will enable you test transaction mining, pending and confirmation logic separately. You can also pass the miner in as a dependency via the constructor, allowing you to mock it for tests.

Reply:
Done in 5180b2c4b8cdedf771a14dfd8ba3bb7f0cfe595b as EthereumTransactionMiner, it does indeed seem cleaner (and the same logic might be useful in other responders as well).

I did not yet add tests for EthereumTransactionMiner, I'll make a separate ticket (it still passes all the integration tests of the responder).